yktrace/src/sir.rs

```diff
@@ -1,92 +1,148 @@
-                            debug_assert!(old.is_none()); // There's one `Types` pack per codegen unit.
-                        }
-                    }
-                }
+                let hdr = read_hdr(&mut curs);
+                let hdr_size = usize::try_from(curs.seek(SeekFrom::Current(0)).unwrap()).unwrap();
+                hdrs.insert(
+                    hdr.cgu_hash,
+                    (sec.name().unwrap().to_owned(), hdr, hdr_size)
+                );
             }
         }
+        Ok(Self { hdrs, exe_obj })
+    }
 
-        Ok(Sir { bodies, types })
+    fn cursor_for_section(&self, sec_name: &str) -> Cursor<&[u8]> {
+        let sec = self.exe_obj.section_by_name(&sec_name).unwrap();
+        Cursor::new(sec.data().unwrap())
     }
-}
 
-impl Display for Sir {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        for body in self.bodies.values() {
-            writeln!(f, "{}", body)?;
+    /// Decode a type in a named section, at an absolute offset from the beginning of that section.
+    fn decode_ty(&self, sec_name: &str, off: SirOffset) -> ykpack::Ty {
+        let mut curs = self.cursor_for_section(&sec_name);
+        curs.seek(SeekFrom::Start(u64::try_from(off).unwrap()))
+            .unwrap();
+        let mut dec = Decoder::from(&mut curs);
+        if let Ok(Some(Pack::Type(t))) = dec.next() {
+            t
+        } else {
+            panic!("Failed to deserialize SIR type");
+        }
+    }
+
+    /// Get the type data for the given type ID.
+    pub fn ty(&self, tyid: &ykpack::TypeId) -> ykpack::Ty {
+        let (cgu, tidx) = tyid;
+        let (ref sec_name, ref hdr, hdr_size) = SIR.hdrs[cgu];
+        let off = hdr.types[usize::try_from(*tidx).unwrap()];
+        self.decode_ty(sec_name, hdr_size + off)
+    }
+
+    /// Decode a body in a named section, at an absolute offset from the beginning of that section.
+    fn decode_body(&self, sec_name: &str, off: SirOffset) -> ykpack::Body {
+        let mut curs = self.cursor_for_section(&sec_name);
+        curs.seek(SeekFrom::Start(u64::try_from(off).unwrap()))
+            .unwrap();
+        let mut dec = Decoder::from(&mut curs);
+        if let Ok(Some(Pack::Body(body))) = dec.next() {
+            body
+        } else {
+            panic!("Failed to deserialize SIR body");
         }
+    }
 
-        for (cgu_hash, types) in self.types.iter() {
-            writeln!(f, "TYPES OF {}", cgu_hash)?;
-            for ty in types {
-                writeln!(f, "{}", ty)?;
+    /// Get the body data for the given symbol name.
+    /// Returns None if not found.
+    pub fn body(&self, body_sym: &str) -> Option<Body> {
+        for (sec_name, hdr, hdr_size) in SIR.hdrs.values() {
+            if let Some(off) = hdr.bodies.get(body_sym) {
+                return Some(self.decode_body(sec_name, hdr_size + off));
             }
         }
+        None
+    }
+}
 
+impl<'m> Display for Sir<'m> {
```

Comment:
General philosophical point: should we be implementing `Display` or `Debug` for these things? I tend to think the latter but I don't feel strongly about it.

Reply:
The convention we've used so is `Display` for pretty-printing, and  `Debug` for everything else.

Reply:
They're both for pretty-printing of sorts, but `Debug` is intended for internal debugging and `Display` for user-facing stuff. Not something to fix in this PR, maybe, but something to think about.

Reply:
> They're both for pretty-printing of sorts

I suppose.

Our `Display` implementations try to concise at the cost of not printing everything like `Debug` would. Also our `Display` stuff for SIR is somewhat "user" facing, as `ykview` prints them.